✨ feat : 온보딩 단계 관리 및 분기 로직/상태 관리 구현

sooyeoneo opened the pull request:
## ☀️ 작업 목적

* 회원 온보딩 시 **단계별 진행 상태를 Enum/Entity 기반**으로 관리
* 선호 지역 특징 선택 및 **적성 검사 분기 로직(인지 여부/수동 선택/AI 추천)** 구현
* **온보딩 이어하기/상태 조회 API** 제공으로 UX 개선
* `RegionType`, `OnboardingStep` Enum 도입으로 **타입 안정성 강화**

---

## 📄 주요 작업 내역

### 1. 단계 관리 & Entity

* `OnboardingStep` Enum : 6단계 온보딩 플로우 정의
* `OnboardingProgress` Entity : 세션 ID 기반 상태 저장
* 적성 인지 여부 분기 로직 포함 (`APTITUDE_CHECK` → MANUAL / AI\_TEST)

---

### 2. DTO

* `OnboardingStartRequestDto`, `OnboardingRegionRequestDto`, `OnboardingAptitudeCheckRequestDto`
* `OnboardingProgressResponseDto` : NextAction 포함 (프론트 라우팅 지원)
* `OnboardingStatusResponseDto` : 진행 상태 조회
* DTO를 **Record 타입으로 변환**하여 불변성 보장

---

### 3. Repository

* `OnboardingProgressRepository` : 회원/세션별 진행 상태 조회, 완료 여부 확인
* 불필요 메소드(`findBySessionId()`, `isOnboardingCompletedByMemberId()`) 삭제

---

### 4. 서비스 로직

* `startOnboarding()` : 온보딩 시작/재개 처리
* `selectRegion()` : 특징 기반 지역 선택 (`URBAN / BALANCED / NATURE`)
* `checkAptitudeKnowledge()` : 적성 인지 여부 확인 후 분기
* `completeOnboarding()` : 완료 처리
* `getOnboardingStatus()`, `getCurrentProgress()` : 이어하기 및 상태 조회
* 세션 ID 생성 규칙 : `ONB-{memberId}-{uuid}`

---

### 5. Controller/API

* `POST /onboarding/start` : 온보딩 시작
* `POST /onboarding/region` : 지역 특징 선택
* `POST /onboarding/aptitude-check` : 적성 인지 여부 확인
* `POST /onboarding/complete` : 온보딩 완료
* `GET /onboarding/status` : 진행 상태 조회 (이어하기용)
* `GET /onboarding/progress` : 현재 단계 상세 조회

---

### 6. 예외 처리

* `ONBOARDING_NOT_STARTED` : 시작하지 않은 상태
* `INVALID_ONBOARDING_STEP` : 잘못된 단계 접근
* `NO_INCOMPLETE_TEST` : 진행 중인 테스트 없음

---

### 7. 테스트

* 온보딩 전체 플로우 시나리오

  * 시나리오 A : 적성을 아는 사용자 (수동 선택)
  * 시나리오 B : 적성을 모르는 사용자 (AI 검사 분기)
* 이어하기/예외 케이스 검증
* 지역 선택 로직 `regionType` 기준으로 업데이트
* 수동 적성 선택 API → Request Body 방식 통일

---

### 8. 리팩토링

* `RegionType` Enum 도입 → 지역 선택 로직 타입 안정성 확보
* DTO Record 변환 및 불필요 코드 제거
* API/로깅 개선 → RESTful 원칙 준수

---

## 📌 참고

* **개선 효과**

  * 온보딩 전체 프로세스를 **단계/Enum 기반 중앙 관리**
  * 지역 선택/적성 검사 분기 구조화 → 가독성과 유지보수성 향상
  * 이어하기 및 상태 조회 기능 → 사용자 경험 강화
* **추가 TODO (#156)**

  * `UserPreferredRegions` Entity 구현
  * 특징별 지역 자동 매핑 로직 추가

---

**Resolves**: #63
**Related to**: #156
**Epic**: #55

---

**리뷰 및 피드백 환영합니다! 🤗**

* #174 PR에 달린 코멘트는 다음 번 PR에 모두 반영하여 수정하도록 하겠습니다!

sooyeoneo committed:
✨ feat : 온보딩 단계 관리 Enum 및 Entity 구현

- OnboardingStep Enum: 6단계 온보딩 플로우 정의
- OnboardingProgress Entity: 진행 상태 저장
- 적성 인지 여부에 따른 분기 로직 포함
- 세션 ID 기반 상태 관리

Resolves: #63

sooyeoneo committed:
✨ feat : 온보딩 요청/응답 DTO 구현

- OnboardingStartRequestDto: 시작 요청
- OnboardingRegionRequestDto: 지역 선택
- OnboardingAptitudeCheckRequestDto: 적성 인지 확인
- OnboardingProgressResponseDto: 진행 상태 응답
- NextAction 정보로 프론트엔드 라우팅 지원

Resolves: #63

sooyeoneo committed:
✨ feat : OnboardingProgressRepository 구현

- 회원별/세션별 온보딩 상태 조회
- 온보딩 완료 여부 확인 쿼리
- JPA 기반 데이터 접근 계층

Resolves: #63

sooyeoneo committed:
✨ feat : 온보딩 비즈니스 로직 구현

- 온보딩 시작/재개 처리
- 선호 지역 선택 로직
- 적성 인지 여부 확인 및 분기 처리
- 온보딩 완료 프로세스
- 트랜잭션 기반 상태 관리

Resolves: #63

sooyeoneo committed:
✨ feat : 온보딩 REST API 엔드포인트 구현

- POST /api/onboarding/start: 온보딩 시작
- POST /api/onboarding/region: 지역 선택
- POST /api/onboarding/aptitude-check: 적성 확인
- POST /api/onboarding/complete: 완료 처리
- GET /api/onboarding/progress: 상태 조회

Resolves: #63

sooyeoneo committed:
🔧 add : 온보딩 관련 ErrorCode 추가

- ONBOARDING_NOT_STARTED: 시작 안됨
- INVALID_ONBOARDING_STEP: 잘못된 단계
- NO_INCOMPLETE_TEST: 진행 중인 테스트 없음

Resolves: #63

sooyeoneo committed:
✅ test : 온보딩 플로우 HTTP 테스트 추가

- 전체 온보딩 플로우 테스트
- 시나리오 A: 적성 아는 사용자
- 시나리오 B: 적성 모르는 사용자
- 예외 케이스 테스트
- 플로우 다이어그램 문서화

Resolves: #63

sooyeoneo committed:
♻️ refactor : 온보딩 지역 선택을 특징 기반으로 변경

[변경사항]
- RegionType Enum 구현: String selectedRegion → RegionType regionType
- OnboardingProgress Entity에 RegionType enum 적용
- OnboardingRegionRequestDto: regionType 필드로 변경
- Service 메서드 시그니처 업데이트
- 지역 특징(도시형/균형형/자연형) 선택 방식으로 수정

[개선효과]
- 지역 선택을 RegionType enum으로 타입 안전성 향상
- 컴파일 타임 검증 강화
- 유효하지 않은 지역 타입 입력 방지

[TODO] (Issue #156)
- UserPreferredRegions Entity 구현
- 특징별 지역 자동 매핑 로직 구현

Related to #156

sooyeoneo committed:
✅ test : 온보딩 HTTP 테스트를 regionType 기준으로 Update

[변경사항]
- region → regionType 필드로 변경
- "서울", "부산" 등 → "URBAN", "BALANCED", "NATURE"
- 대소문자 변환 테스트 추가 (urban, Urban)
- console.log 출력 필드 일치 (data.regionType)

[테스트 시나리오]
- 도시형/균형형/자연형 선택 테스트
- 대소문자 처리 검증
- 시나리오별 플로우 업데이트

Related to #63, #156

sooyeoneo committed:
Merge branch 'develop' into feature/onboarding

# Conflicts:
#	src/main/java/com/salayo/locallifebackend/global/error/ErrorCode.java

sooyeoneo committed:
🔥 remove : ErrorCode에서 NO_INCOMPLETE_TEST 예외 중복 삭제

sooyeoneo committed:
Merge branch 'develop' into feature/onboarding

sooyeoneo committed:
🔧 add : 온보딩 진행 상태 관리를 위한 DTO 및 Enum 개선

[변경 사항]
- OnboardingProgressResponseDto 타입 안정성 강화
  - regionType: String → RegionType enum으로 변경
  - NextAction.type: String → OnboardingStep enum으로 변경
  - TODO 주석 추가: Issue #156 지역 매핑 구현 예정

- OnboardingStep enum 기능 확장
  - getNextStep() 메소드 추가: 적성 인지 여부에 따른 분기 로직
  - switch expression 활용한 다음 단계 결정 로직
    - APTITUDE_CHECK → 적성 인지 시 MANUAL, 모를 시 AI_TEST

[개선 효과]
- Enum 활용으로 타입 안정성 및 가독성 향상
- 온보딩 단계 전환 로직 중앙화
- 컴파일 타임에 타입 체크로 런타임 오류 방지

Refs: #63

sooyeoneo committed:
🔥 remove : OnboardingStartRequestDto 삭제

sooyeoneo committed:
♻️ refactor : 온보딩 DTO를 Record로 변환 및 상태 조회 DTO 추가

[변경 사항]
- OnboardingRegionRequestDto: class → record 변환
  - 불변성 보장 및 보일러플레이트 코드 제거
  - TODO 주석: Issue #156 지역 매핑 구현 예정

- OnboardingAptitudeCheckRequestDto: class → record 변환
  - @NotNull 검증 유지
  - knowsAptitude 필드로 적성 인지 여부 표현

- OnboardingStatusResponseDto 신규 추가
  - 온보딩 진행 상태 조회용 DTO
  - 정적 팩토리 메서드: complete(), inProgress()
  - @JsonInclude로 null 필드 제외

[개선 효과]
- Record 활용으로 코드 간결화 및 불변성 보장
- 온보딩 상태 조회 기능 추가 (이어하기 지원)

Refs: #63 #156

sooyeoneo committed:
✨ feat : 온보딩 서비스 단계별 진행 로직 및 이어하기 기능 구현

[구현 기능]
- startOnboarding(): 온보딩 시작 및 기존 진행 상태 확인
  - 이미 완료된 경우 완료 응답 반환
  - 진행 중인 경우 이어하기 가능

- selectRegion(): 선호 지역 특징 선택
  - URBAN(도시형), BALANCED(균형형), NATURE(자연형)
  - TODO: Issue #156 지역 자동 매핑 예정

- checkAptitudeKnowledge(): 적성 인지 여부 확인 및 분기
  - true: 수동 선택 화면으로 이동
  - false: AI 적성 검사로 이동

- getOnboardingStatus: 미완료 온보딩 조회 (이어하기용)

- getCurrentProgress(): 현재 진행 상태 조회
  - 온보딩 이어하기를 위한 상태 정보 제공

[개선 사항]
- 단계별 검증 로직 강화
- 세션 ID 생성 로직 추가 (ONB-{memberId}-{uuid})
- 트랜잭션 처리 최적화

Refs: #63

sooyeoneo committed:
♻️ refactor : OnboardingController API 개선

[변경 사항]
- POST /onboarding/start: 온보딩 시작
- POST /onboarding/region: 선호 지역 선택
- POST /onboarding/aptitude-check: 적성 인지 여부 확인
- POST /onboarding/complete: 온보딩 완료
- GET /onboarding/status: 진행 상태 조회 (이어하기용)
- GET /onboarding/progress: 현재 단계 상세 조회

[개선 효과]
- RESTful API 설계 원칙 준수
- 온보딩 이어하기 기능 지원
- 로깅 추가로 디버깅 용이성 향상

Refs: #63

sooyeoneo committed:
🔥 remove : 불필요한 메서드 제거

- OnboardingProgressRepository
  - findBySessionId(), isOnboardingCompletedByMemberId() 삭제

sooyeoneo committed:
✅ test : 온보딩 HTTP 테스트 - 수동 적성 선택 API Request Body 방식으로 통일

- 수동 적성 선택 API 호출 방식 변경
  - Query Parameter 방식 제거 (?aptitudeType=XXX)
  - 모든 적성 타입을 Request Body로 통일
  - AptitudeManualSelectRequestDto (Record) 형식에 맞춤

- 5개 적성 타입 모두 동일한 형식으로 통일
  - NATURE, HISTORY_CULTURE, ART_CREATION, COMMUNITY, TECH
  - JSON body: { "aptitudeType": "XXX" }

Refs: #63

jiyoon0000 commented on src/main/java/com/salayo/locallifebackend/domain/onboarding/entity/OnboardingProgress.java:
```diff
@@ -0,0 +1,98 @@
+    private OnboardingStep currentStep;
+    
+    @Column(name = "is_completed", nullable = false)
+    private Boolean isCompleted;
```
요거도 원시타입 boolean으로 수정하시면 좋을 것 같습니다!

sooyeoneo replied:
넵 수정하였습니다!

jiyoon0000 commented on src/main/java/com/salayo/locallifebackend/domain/onboarding/dto/OnboardingProgressResponseDto.java:
```diff
@@ -0,0 +1,156 @@
+	private final Boolean isCompleted;
+	private final RegionType regionType;  // String에서 RegionType으로 변경
+	private final Boolean knowsAptitude;
+	private final String message;
```
이거도 전 pr에서 얘기했듯이 조금 더 명확한 네이밍을 쓰면 좋을 것 같습니다😎

sooyeoneo replied:
확인해주셔서 감사합니다!

jiyoon0000 commented on src/main/java/com/salayo/locallifebackend/domain/onboarding/service/OnboardingService.java:
```diff
@@ -0,0 +1,205 @@
+     * 온보딩 진행 상태 조회 (내부 사용)
+     */
+    private OnboardingProgress getOnboardingProgress(Long memberId) {
+        return onboardingProgressRepository.findByMemberId(memberId)
```
온보딩은 이미 로그인된 사용자의 상태를 다루기 때문에 항상 인증된 Member 객체가 먼저 조회됩니다.
따라서 MemberId를 불러오는 것 보다는 findByMember를 사용하는 것이 조금 더 자연스러울 것 같아요

그렇게 되면 OnbordingProgressRepository에 있는 메서드는 삭제해도 될 것 같아요!
```
    @Query("SELECT op FROM OnboardingProgress op WHERE op.member.id = :memberId")
    Optional<OnboardingProgress> findByMemberId(@Param("memberId") Long memberId);
```

sooyeoneo replied:
넵 다 수정하였습니다! 감사합니다🥹💗

jiyoon0000 commented on src/main/resources/http/onboarding.http:
```diff
@@ -0,0 +1,363 @@
+### ========================================
+
+### 온보딩 완료 처리
+POST {{host}}/api/onboarding/complete
```
260, 284번 줄에서 요청 URL에  /api 가 붙어있는데, controller 경로와 다른 것 같아요!! 수정 부탁드려요🥹

sooyeoneo replied:
헙 아직 남아있었네요 수정하였습니다!😂

sooyeoneo committed:
♻️ refactor : OnboardingProgress 엔티티 타입 최적화 - isCompleted 원시타입, knowsAptitude 래퍼타입

[변경 사항]
- isCompleted: boolean 원시 타입 유지
  - 항상 true/false 값을 가짐
  - NPE 위험 없음

- knowsAptitude: Boolean 래퍼 타입 유지
  - null: 아직 질문하지 않음 (초기 상태)
  - true/false: 적성 인지 여부 응답

- Builder 및 메서드 타입 정리
  - updateKnowsAptitude(Boolean) 파라미터 타입 수정
  - 온보딩 시작 시 knowsAptitude null로 초기화

[개선 효과]
- 각 필드의 의미에 맞는 타입 사용
- null 상태가 필요한 곳에만 래퍼 타입 적용
- 타입 안정성과 유연성 균형

Refs: #63

j-hann commented on src/main/java/com/salayo/locallifebackend/domain/onboarding/service/OnboardingService.java:
```diff
@@ -0,0 +1,205 @@
+    /**
+     * 세션 ID 생성
+     */
+    private String generateSessionId(Long memberId) {
+        return String.format("ONB-%d-%s", memberId, UUID.randomUUID().toString().substring(0, 8));
```
수연님 혹시 세션 id 생성하실 때  memberId를 넣어서 생성하시는 이유가 있을까요?

`OnboardingProgressResponseDto`, `OnboardingStatusResponseDto`에 해당 필드가 클라이언트에 노출되는 것으로 보이는데 memberId를 포함하여 생성하는 것이 보안적으로 좋지 않을 것 같다는 의견입니다!

그리고 전에 주문번호 생성 시 제안해주셨던 방식처럼 조금 더 긴 문자열로 생성하면 어떨지 제안드립니다!

sooyeoneo replied:
네! 맞습니다. 지금 현재는 임시로 해둔거라서, 온보딩 거의 완성되고 합치는 과정에서 
ai-aptitude 쪽과 onboarding에서 세션id 생성하는 방식을 다시 정리를 해야할 것 같습니다! 
확인 감사합니다! ☺️

sooyeoneo committed:
♻️ refactor : DTO 필드명을 더 명확하게 변경 (message → guideMessage)

[변경사항]
- OnboardingProgressResponseDto: message → guideMessage
- OnboardingProgressResponseDto.builder(): 필드명 일관성 적용

[변경 이유]
- CommonResponseDto.message와 혼동 방지
- 필드의 구체적 역할 명시 (온보딩 진행 가이드 안내 메시지)

[개선효과]
- API 응답 구조 명확화
- 프론트엔드 개발 시 혼동 방지
- 코드 가독성 향상

Resolves: #174

sooyeoneo committed:
♻️ refactor : OnboardingService에서 Member 객체 직접 사용하도록 개선

## 변경 사항
### OnboardingService
- getOnboardingProgress() 메서드 파라미터 변경: Long memberId → Member member
- 모든 public 메서드에서 Member 객체 먼저 조회 후 전달
  - selectRegion(), checkAptitudeKnowledge(), completeOnboarding(), getCurrentProgress()
- getOnboardingStatus() 버그 수정: findByMemberId → findByMember 사용

### OnboardingProgressRepository
- findByMemberId() JPQL 쿼리 메서드 제거
- findByMember(Member member) 사용 (Spring Data JPA 자동 구현)

## 개선 효과
- 객체지향 설계 원칙 준수: 원시 타입(Long) 대신 도메인 객체(Member) 사용
- 타입 안정성 향상: 컴파일 시점 타입 검증으로 실수 방지
- 코드 명확성: 비즈니스 의미가 명확하게 표현됨
- Repository 단순화: 불필요한 JPQL 제거

## 버그 수정
- getOnboardingStatus()에서 잘못된 findById 사용 수정

Resolves: #174

j-hann reviewed (approved):
굿굿👍
